include/circt/Dialect/ESI/ESIPasses.td

```diff
@@ -54,4 +54,11 @@ def LowerESItoHW: Pass<"lower-esi-to-hw", "mlir::ModuleOp"> {
   let dependentDialects = ["circt::comb::CombDialect", "circt::hw::HWDialect"];
 }
 
+def ESIWrapCombinational: Pass<"esi-wrap-combinational", "mlir::ModuleOp"> {
+  let summary = "Wrap around each hw.external.module with EsiWrap attribute set to combinational";
```

Comment:
I still don't understand what this pass does. Can you add a test and a description? What do you mean by "set it to combinational"?

It doesn't look like it, but is this just to for ESI not to use SV interfaces on extern modules?

Reply:
I will add a better description to the Pass.
This pass enables us to embed an arbitrary purely combinational module (with no notion of valid ready) into a handshaking environment. Here is an example:
We start with a handshake program. As you can see the hw.module @ext_add is purely combinational. We now want to instantiate this module in our handshake program, which we do with the handshake.external_instance instruction. Instead of referencing a handshake.func we now reference a hw.module. (Note: probably this handshake.external instance is not needed since we can do it directly with an empty handshake_function see comment from Morten)
```mlir
module {
  hw.module @ext_add(%a : i32, %b : i32) -> (out: i32) {
    %0 = comb.add %a, %b : i32
    hw.output %0 : i32
  }
  handshake.func @main(%a: i32, %b: i32, %ctrl : none) -> (i32, none) {
    // Instantiate ext_add
    %0 = handshake.external_instance @ext_add(%a, %b) : (i32, i32) -> (i32)
    handshake.return %0, %ctrl : i32, none
  }
}
```

From here we lower it to HW/ESI with `--lower-handshake-to-hw`. This looks like this. Important to note here is that hw.module has still no notion of handshaking. But the handshake_external_instance_ext_add module will have a handshaking logic, which instantiates the hw.ext_add instance and adds the necessary wrapping logic around it.
```mlir
module {
  hw.module.extern @handshake_external_instance_ext_add(%in0: !esi.channel<i32>, %in1: !esi.channel<i32>) -> (out0: !esi.channel<i32>) attributes {esiWrap = {target = @ext_add, type = "combinational"}}
  hw.module @ext_add(%a: i32, %b: i32) -> (out: i32) {
    %0 = comb.add %a, %b : i32
    hw.output %0 : i32
  }
  hw.module @main(%a: !esi.channel<i32>, %b: !esi.channel<i32>, %ctrl: !esi.channel<i0>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i0>) {
    %handshake_external_instance0.out0 = hw.instance "handshake_external_instance0" @handshake_external_instance_ext_add(in0: %a: !esi.channel<i32>, in1: %b: !esi.channel<i32>) -> (out0: !esi.channel<i32>)
    hw.output %handshake_external_instance0.out0, %ctrl : !esi.channel<i32>, !esi.channel<i0>
  }
}
```

With the `--esi-wrap-combinational` pass we now materialize this hw.module.external and add all the wrapping and handshaking logic. This then looks like this:
```mlir
module {
  hw.module @handshake_external_instance_ext_add(%in0: !esi.channel<i32>, %in1: !esi.channel<i32>) -> (out0: !esi.channel<i32>) {
    %rawOutput, %valid = esi.unwrap.vr %in0, %0 : i32
    %rawOutput_0, %valid_1 = esi.unwrap.vr %in1, %0 : i32
    %ext_add.out = hw.instance "ext_add" @ext_add(a: %rawOutput: i32, b: %rawOutput_0: i32) -> (out: i32)
    %chanOutput, %ready = esi.wrap.vr %ext_add.out, %0 : i32
    %0 = comb.and %valid, %valid_1, %ready : i1
    hw.output %chanOutput : !esi.channel<i32>
  }
  hw.module @ext_add(%a: i32, %b: i32) -> (out: i32) {
    %0 = comb.add %a, %b : i32
    hw.output %0 : i32
  }
  hw.module @main(%a: !esi.channel<i32>, %b: !esi.channel<i32>, %ctrl: !esi.channel<i0>, %clock: i1, %reset: i1) -> (out0: !esi.channel<i32>, out1: !esi.channel<i0>) {
    %handshake_external_instance0.out0 = hw.instance "handshake_external_instance0" @handshake_external_instance_ext_add(in0: %a: !esi.channel<i32>, in1: %b: !esi.channel<i32>) -> (out0: !esi.channel<i32>)
    hw.output %handshake_external_instance0.out0, %ctrl : !esi.channel<i32>, !esi.channel<i0>
  }
}
```
From here CIRCT will take over. 

So why do we want something like this: In the combinational case we could have a latency benefit but with a reduced wall clock frequency due to longer combinational paths. It gets more interesting if we take the same idea and apply it to a statically pipelined module. (I am now working on such a wrapping pass). In this case we could achieve almost the same performance but with potentially area usage savings.

Reply:
> This pass enables us to embed an arbitrary purely combinational module (with no notion of valid ready) into a handshaking environment.

That makes sense. Since it sounds like a relatively specific use case, I would be more comfortable with this pass being in the handshake dialect then. Also, is using a `hw.module` really appropriate? Since you need to guarantee that the logic is entirely combinational, wouldn't an operation which has that verifier be more appropriate?

> So why do we want something like this: In the combinational case we could have a latency benefit but with a reduced wall clock frequency due to longer combinational paths. It gets more interesting if we take the same idea and apply it to a statically pipelined module. (I am now working on such a wrapping pass). In this case we could achieve almost the same performance but with potentially area usage savings

Take a look at the (relatively new) [`pipeline.pipeline` op](https://circt.llvm.org/docs/Dialects/Pipeline/#pipelinepipeline-circtpipelinepipelineop). It was intended for exactly this. We haven't used it yet for that purpose (or at all) so it may not fit exactly, but I would encourage you to extend/change that instead of re-inventing something.

Coincidentally, the pipeline ops aren't intended to support any state (they start as combinational only) and the state is scheduled. I've always been uncomfortable with the "combinational only" starting condition being implicit. I'd much rather start with a combinational container op being scheduled during a lowering pass into a pipelined operation.